# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **

I think this is a Classification problem, we learned that in Regression problems we want to predict continuoues values, like the price of a House for example, in this case we need to indentify students who may need intervention before they fail to graduate, so we need to classiffy the students according to this premise.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [9]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [10]:
# TODO: Calculate number of students
n_students = student_data['school'].count()

# TODO: Calculate number of features
n_features = len(student_data.columns[:-1])

# TODO: Calculate passing students
n_passed = student_data.loc[student_data['passed'] == 'yes']["passed"].count()

# TODO: Calculate failing students
n_failed = student_data.loc[student_data['passed'] == 'no']["passed"].count()

# TODO: Calculate graduation rate
grad_rate = (n_passed*100)/float(n_students)

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [11]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [12]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [13]:
# TODO: Import any additional functionality you may need here
from sklearn.model_selection import train_test_split

# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, random_state=42)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB) **Nope**
- Decision Trees **(Nope)**
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting) ** Gradient Boosting **
- K-Nearest Neighbors (KNeighbors) **(Nope)**
- Stochastic Gradient Descent (SGDC) **(Nope)**
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **

I chose the following:

1) Gradient Boosting.
    
**Describe one real-world application **
    
    A particular implementation of gradient boosting, XGBoost
    "is consistently used to win machine learning competitions on Kaggle"
     

**What are the strengths of the model; when does it perform well?**

- Natural handling of data of mixed type (= heterogeneous features)
- Predictive power
- Robustness to outliers in output space (via robust loss functions)

**What are the weaknesses of the model; when does it perform poorly?**
- Scalability, due to the sequential nature of boosting it can hardly be parallelized.

**What makes this model a good candidate for the problem, given what you know about the data?**
- In the lessons we learned that this method of boosting is going to get better and better over time, so I just want to see if that is true or at least see if that applys here, also in the comparations that I made I see this mehtod gave me better results.

** Refferences**

http://scikit-learn.org/stable/modules/ensemble.html#gradient-boosting

http://blog.kaggle.com/2017/01/23/a-kaggle-master-explains-gradient-boosting/

https://github.com/dmlc/xgboost
 
2) Support Vector Machines (SVM)

**Describe one real-world application **

Cancer Classificaion

**What are the strengths of the model; when does it perform well?**

- Flexibility in choosing a similarity function
- Sparseness of solution when dealing with large data sets
- Ability to handle large feature spaces
- Overfitting can be controlled by soft margin approach
- Feature Selection

**What are the weaknesses of the model; when does it perform poorly?**

- It is sensitive to noise
- If the number of features is much greater than the number of samples, the method is likely to give poor performances.
- SVMs do not directly provide probability estimates, these are calculated using an expensive five-fold cross-validation

**What makes this model a good candidate for the problem, given what you know about the data?**

I think it is a good match with the data we have because with svm we need to fin a boundary of all the students and find which of them are suppossed to pass the final exam and graduate, I think that with this model we can achieve this, also another advantage I see is that has a good resources usage and this is another requirement in the problem.

** Refferences**

www.iro.umontreal.ca/~pift6080/H09/documents/papers/svm_tutorial.ppt

http://scikit-learn.org/stable/modules/svm.html

3) Logistic Regression

    
**Describe one real-world application **

Healthcare : Analyzing a group of over million people for myocardial infarction within a period of 10 years is an application area of logistic regression.

**What are the strengths of the model; when does it perform well?**
- it is more robust: the independent variables don’t have to be normally distributed, or have equal variance in each group
- Is better when you are dealing with very high dimension data.
 
**What are the weaknesses of the model; when does it perform poorly?**
- requires large sample size to achieve stable results.

**What makes this model a good candidate for the problem, given what you know about the data?**
Logistic Regression I think is a good candidate because will help us to define a binary outcome given the fetures that we have, 0 or 1 on if the student will pass the final exam or not.

**Refferences**

https://www.quora.com/What-are-applications-of-linear-and-logistic-regression

https://victorfang.wordpress.com/2011/05/10/advantages-and-disadvantages-of-logistic-regression/



### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [14]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [10]:
# TODO: Import the three supervised learning models from sklearn
# from sklearn import model_A
from sklearn.svm import SVC
from sklearn import linear_model
from sklearn.ensemble import GradientBoostingClassifier

# from sklearn import model_B
# from sklearn import model_C

# TODO: Initialize the three models
clf_A = GradientBoostingClassifier(random_state=42)
clf_B = linear_model.LogisticRegression(random_state=42)
clf_C = SVC(random_state=42)

# TODO: Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# TODO: Execute the 'train_predict' function for each classifier and each training set size
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
print("######")
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
print("######")
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
print("######")
train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
print("######")
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
print("######")
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
print("######")
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
print("######")
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
print("######")
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)
print("######")
print("######")
train_predict(clf_B, X_train, y_train, X_test, y_test)

Training a GradientBoostingClassifier using a training set size of 100. . .
Trained model in 0.0540 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7519.
######
Training a GradientBoostingClassifier using a training set size of 200. . .
Trained model in 0.0760 seconds
Made predictions in 0.0160 seconds.
F1 score for training set: 0.9964.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7591.
######
Training a GradientBoostingClassifier using a training set size of 300. . .
Trained model in 0.1070 seconds
Made predictions in 0.0010 seconds.
F1 score for training set: 0.9739.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7794.
######
Training a LogisticRegression using a training set size of 100. . .
Trained model in 0.0000 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.8593.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7647.
###

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - Gradient Boosting?**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0510 seconds      |    0.0000 seconds      |      1.0000      |    0.7519       |
| 200               |     0.0810 seconds      |    0.0000 seconds      |      0.9964      |    0.7591       |
| 300               |     0.1020 seconds      |    0.0020 seconds      |      0.9739      |    0.7794       |

** Classifer 2 - Logistic Regression?**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0000 seconds      |    0.0000 seconds      |      0.8593      |    0.7647       |
| 200               |     0.0000 seconds      |    0.0000 seconds      |      0.8562      |    0.7914       |
| 300               |     0.0190 seconds      |    0.0000 seconds.     |      0.8468      |    0.8060       |

** Classifer 3 - Support Vector Machines (SVM)?**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0000 seconds      |    0.0000 seconds      |      0.8777      |    0.7746       |
| 200               |     0.0000 seconds      |    0.0150 seconds      |      0.8679      |    0.7815       |
| 300               |     0.0000 seconds      |    0.0160 seconds.     |      0.8761      |    0.7838       |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **

** Selected Model **

Based on the results obtained earlier in the experiments I selected Logistic Regression.

**Which model is generally the most appropriate based on the available data, limited resources, cost, and performance**

Bases on what I researched and the results I think Logistic Regression suit better to the data we have, in performance is verry effective compared to the other models, the training time is much less than the Gradient Boosting Model prediction time and a little bit better than SVM, the prediction time was really close in all 3 Models but the logistic regression was the second lowest, regarding the F1 Scores I found that the first on Gradient Boosting have the best Score for training with values closest to 1, but in the F1 Score for testing was the lowest, this let me think that the Gradient Boosting tends to overfit in with this data, the best Testing Score was to Logistic Regression by a little margin, but I think is the best one although the SVM Model I think came very close.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **

We have a Classification problem for which we selected a Model that suits the need for this problem, we selected Logistic Regression because we need a discrete answer for the question "Will the student pass the final exam?", Yes or No, 0 or 1, if the answer to this question is "Yes" then we need to do an intervention for the student. Logistic Regression let us do exactly this, by making binary predictions given a certain number of independend features.

Firs we take all the historical data and train the Logistic Regression Model, this model takes historical information about students, like (Age, Gender, etc) and creates a model that assings weights to every feature giving an importance to ech one of them, we also have a historic of those students, they passed or not?, the after the training when we want to predict an outcome for a new student , we take the new student's features use the model to predict the outcome based on the function used by the Model called (sigmoid function), then we apply the values to this function and obtain the final probability that give us the answer of the student, passed final exam Yes or No the.


### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [9]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import f1_score
from sklearn.metrics import make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {
    'C': np.logspace(-4,4,17),
    'penalty':['l1', 'l2'],
    'class_weight':[None, 'balanced'],
    }

# TODO: Initialize the classifier
clf = linear_model.LogisticRegression(random_state=42)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label="yes")

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, cv=5, scoring=f1_scorer, 
                        verbose=1, n_jobs=-1)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Fitting 5 folds for each of 68 candidates, totalling 340 fits


[Parallel(n_jobs=-1)]: Done  34 tasks      | elapsed:    3.8s


Made predictions in 0.0000 seconds.
Tuned model has a training F1 score of 0.8323.
Made predictions in 0.0000 seconds.
Tuned model has a testing F1 score of 0.7891.


[Parallel(n_jobs=-1)]: Done 340 out of 340 | elapsed:    6.7s finished


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

**What is the final model's F1 score for training and testing?**

** Tuned Classifier**
F1 Score Training: 0.8323
F1 Score Testing: 0.7891

** Untuned Classifier **
F1 score for training: 0.8468
F1 score for test: 0.8060

**How does that score compare to the untuned model?**

Comparing the results I got that the training score of the tuned version is a bit lower than the untunser classifier, and the testing score of the tuned version is a bit lower than the untuned version.

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.